aiida/orm/nodes/data/array/cube.py

```diff
@@ -0,0 +1,246 @@
+
+    @atomic_numbers.setter
+    def atomic_numbers(self, atomic_numbers):
+        self.set_array('atomic_numbers', atomic_numbers)
+
+    @property
+    def atomic_charges(self):
+        """Atomic charges in the structure."""
+        return self.get_array('atomic_charges')
+
+    @atomic_charges.setter
+    def atomic_charges(self, atomic_charges):
+        self.set_array('atomic_charges', atomic_charges)
+
+    @property
+    def atomic_coordinates(self):
+        """Atomic coordinates in the structure."""
+        return self.get_array('atomic_coordinates')
+
+    @atomic_coordinates.setter
+    def atomic_coordinates(self, atomic_coordinates):
+        self.set_array('atomic_coordinates', atomic_coordinates)
+
+    @property
+    def data(self):
+        """3D array containing the mesh data."""
+        return self.get_array('data')
+
+    @data.setter
+    def data(self, data):
+        self.set_array('data', data)
+
+    @property
+    def data_units(self):
+        """Units of the mesh data."""
+        return self.get_attribute('data_units')
+
+    @data_units.setter
+    def data_units(self, data_units):
+        self.set_attribute('data_units', data_units)
+
+    def read(self, file, data_units=None):
+        """Parse cube file."""
+
+        # It should be either a string pointing to a file, or a file object.
+        if isinstance(file, str):
+            fiter = iter(open(file).readlines())
+        else:
+            fiter = iter(file.readlines())
+
+        # Comment.
+        self.comment = [next(fiter).rstrip(), next(fiter).rstrip()]
+
+        # Number of atoms and origin.
+        line = next(fiter).split()
+        natoms = int(line[0])
+        self.origin = np.array(line[1:], dtype=float)
+
+        # Data shape and voxel.
+        shape = np.empty(3, dtype=int)
+        voxel = np.empty((3, 3))
+        for i in range(3):
+            nmb, x, y, z = [float(s) for s in next(fiter).split()]
+            shape[i] = int(nmb)
+            voxel[i] = np.array([x, y, z])
+
+        self.voxel = voxel
+
+        # Atomic numbers, charges and positions.
+        numbers = np.empty(natoms, int)
+        charges = np.empty(natoms, int)
+        positions = np.empty((natoms, 3))
+        for i in range(natoms):
+            line = next(fiter).split()
+            numbers[i] = int(line[0])
+            charges[i] = float(line[1])
+            positions[i] = [float(s) for s in line[2:]]
+        self.atomic_numbers = numbers
+        self.atomic_charges = charges
+        self.atomic_coordinates = positions
+
+        # Data mesh.
+        data = np.empty(shape[0] * shape[1] * shape[2], dtype=float)
+        cursor = 0
+        for i, line in enumerate(fiter):
+            l_split = line.split()
+            data[cursor:cursor + len(l_split)] = l_split
+            cursor += len(l_split)
+        self.data = data.reshape(shape)
+
+        if data_units:
+            self.data_units = data_units
+
+    def everything_ok(self):
```

Comment:
Is there a reason you are only checking whether these fields are set and not the values whatsoever? In the setters, the only explicit check is on the comments, to verify they are strings. For all the others, both attributes and arrays, you do not check the value whatsoever. The `set_array` will check that what is passed is a numpy array, so at least the type will be correct, but you could set an empty array, or an array with incommensurate dimensions. The `set_attribute` is worse, because it happily accepts `None`. I am not sure what the usefulness of this validator method is if its checks are very limited. Even though `None` won't be accepted in the constructor, afterwards it is perfectly fine to set it.